lms/services/document_url.py

```diff
@@ -0,0 +1,107 @@
+import re
+from functools import lru_cache
+from typing import Optional
+from urllib.parse import unquote
+
+from lms.services.vitalsource import VitalSourceService
+
+
+class DocumentURLService:
+    """A service for getting LTI launch document URLs."""
+
+    def __init__(self, assignment_service):
+        self._assignment_service = assignment_service
+
+    @lru_cache(1)
+    def get_document_url(self, context, request) -> Optional[str]:
+        """
+        Get the configured document for this LTI launch.
+
+        This will try all known methods for finding out, and will return None
+        if no configuration can be found.
+        """
+        for document_url_source in (
+            self._from_deep_linking_provided_url,
+            self._from_canvas_file,
+            self._from_legacy_vitalsource_book,
+            self._from_assignment_in_db,
+        ):
+            if document_url := document_url_source(context, request):
+                return document_url
+
+        return None
+
+    _ENCODED_URL = re.compile("^(?:https?|canvas|vitalsource)%3a", re.IGNORECASE)
+
+    @classmethod
+    def _from_deep_linking_provided_url(cls, _context, request):
+        """Get the URL from our own configuration set during deep linking."""
+
+        if url := request.params.get("url"):
+            # Work around a bug in Canvas's handling of LTI Launch URLs in
+            # SpeedGrader launches where query params get double-encoded.
+            # See https://github.com/instructure/canvas-lms/issues/1486
+            if cls._ENCODED_URL.match(url):
+                url = unquote(url)
+
+        return url
+
+    @classmethod
+    def _from_canvas_file(cls, context, request):
+        """Get a Canvas file URL based on temporary data passed from Canvas."""
```

Comment:
This is the same as URL, you mean temporary here because we just use it to construct the `canvas://` URL?

Reply:
The data is temporary in the sense that Canvas changes it every call potentially. It's not something we can store. 

I was trying to distil down the essence of the long comment on this from before. Looks like this is too far the other way and could do with some extra clarification.